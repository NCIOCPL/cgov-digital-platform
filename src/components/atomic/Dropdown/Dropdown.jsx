import React from 'react';
import PropTypes from 'prop-types';
import InputLabel from '../InputLabel';
import { uniqueIdForComponent } from '../../../utilities/utilities';

//  Class representing a dropdown
//
//  This component expects that its children will be <options> elements
//
//  Required props:
//  - label: string. Sets the text for the input's label
//  - children: node. Series of <option> elements
//
//  Optional props:
//  - id: string. Sets the select's id attribute nd the label's for attribute
//  - value: string. Sets default select choice. Must match the value of one of the <option> elements. Or if an empty string, a blank placeholder is added.
//  - required: bool. Adds required label, required attribute and aria-required='true'
//  - errorMessage: string. If present triggers the error state and displays the error message

export default class Dropdown extends React.Component {
  //  Constructor
  //  @param {object} props The props. See proptypes below.
  //
  //  Set initial state
  //  value: default selected option passed from a prop
  //  hasError: tracks if the field has an error
  //  errorMessage: displayed message when the field hasError

  static propTypes = {
    id: PropTypes.string,
    label: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
    required: PropTypes.bool,
    hasError: PropTypes.bool,
    errorMessage: PropTypes.string,
    value: PropTypes.string,
    action: PropTypes.func,
  };

  static defaultProps = {
    required: false,
    hasError: false,
    action: () => {},
  };

  constructor(props) {
    super(props);
    this.state = {
      value: this.props.value || '',
      hasError: this.props.errorMessage ? true : false,
      errorMessageBody: this.props.errorMessage
        ? this.props.errorMessage
        : null,
    };
  }

  // check to see if an Id was passed in, if not generate one.
  componentWillMount() {
    this.id = this.props.id ? this.props.id : uniqueIdForComponent(this);
  }

  // If a errorMessage is passed after initial render, adjust the state accordingly
  componentWillReceiveProps({ errorMessage }) {
    if (errorMessage) {
      this.setState({
        hasError: true,
        errorMessageBody: errorMessage,
      });
    }
  }

  //  Update the state when user selects a new option
  //  @param {event} event The async event

  _handleChange(event) {
    this.setState({
      value: event.target.value,
    });
    this.props.action(event.target.value);
  }

  render() {
    let errorMessage = null;
    if (this.state.hasError) {
      errorMessage = (
        <span className="usa-error-message" role="alert">
          {this.state.errorMessageBody}
        </span>
      );
    }

    let emptyPlaceholder = null;

    if (this.state.value === '') {
      emptyPlaceholder = (
        <option disabled value="">
          Select ...
        </option>
      );
    }

    return (
      <React.Fragment>
        <InputLabel
          htmlFor={this.id}
          required={this.props.required}
          label={this.props.label}
        />

        {errorMessage}

        <select
          className="usa-select"
          name={this.props.id}
          id={this.id}
          value={this.state.value}
          required={this.props.required}
          onChange={this._handleChange.bind(this)}
        >
          {emptyPlaceholder}
          {this.props.children}
        </select>
      </React.Fragment>
    );
  }
}
