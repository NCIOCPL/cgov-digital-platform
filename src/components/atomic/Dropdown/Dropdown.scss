.cts-select {
  border-width: 1px;
  border-color: #bdbdc2;
  border-style: solid;
  appearance: none;
  border-radius: 0;
  color: #2e2e2e;
  display: block;
  height: 2.5rem;
  max-width: 30rem;
  padding: 0.5rem;
  width: 100%;
  // 
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxMCAxNCI+PGcgZmlsbD0iIzNkNDU1MSIgZmlsbC1ydWxlPSJldmVub2RkIj48cGF0aCBkPSJNLjYyNjM5NjI4IDUuOTk5OTM4OTZIOS4zNzI3Mzg5Yy4yNTMzOTE3My4wMDEzMDg1Ni40ODI0MTYtLjE1MDc0MjY0LjU3OTU2MjAyLS4zODQ3NzYyNC4wOTcxNDU5OC0uMjM0MDMzNi4wNDMxMjI5NC0uNTAzNTc2MTYtLjEzNjcwOTIyLS42ODIwOTY0TDUuNDQwMTgxNS4xODM1NDM1QzUuMzI0NzU2MDYuMDY2MTMxMTYgNS4xNjcwMDgzIDAgNS4wMDIzNjExNCAwYy0uMTY0NjQ3MTYgMC0uMzIyMzk0OS4wNjYxMzExNy0uNDM3ODIwMzcuMTgzNTQzNUwuMTgzNTQzNSA0LjkzMzA2NjMyQy4wNjYxMzExNiA1LjA0ODQ5MTggMCA1LjIwNjIzOTUzIDAgNS4zNzA4ODY3cy4wNjYxMzExNy4zMjIzOTQ5Mi4xODM1NDM1LjQzNzgyMDM4Yy4xMTQ2NDQ0LjEyMjQ5MzQuMjc1MDgwMDYuMTkxNzcyNDQuNDQyODUyNzguMTkxMjMxODh6TTkuMzcyNzM4OSA3Ljk5OTk2ODVILjYyNjM5NjNjLS4xNjYyNjUzLS4wMDMxNjE4My0uMzI2NDc2MDguMDYyMzc4OTQtLjQ0Mjg1MjguMTgxMTY3MDVDLjA2NjEzMTIgOC4yOTY1NjEgMCA4LjQ1NDMwODc1IDAgOC42MTg5NTU5MnMuMDY2MTMxMTcuMzIyMzk0OTIuMTgzNTQzNS40Mzc4MjAzN2w0LjM4MDk5NzI2IDQuNzU5NjgwMkM0LjY3OTk2NjIyIDEzLjkzMzg2ODggNC44Mzc3MTM5NiAxNCA1LjAwMjM2MTEzIDE0cy4zMjIzOTQ5Mi0uMDY2MTMxMi40Mzc4MjAzOC0uMTgzNTQzNWw0LjM3NTQxMDItNC43NDk2MTUzN2MuMTc5ODMyMTYtLjE3ODUyMDIzLjIzMzg1NTItLjQ0ODA2Mjc4LjEzNjcwOTIyLS42ODIwOTYzOC0uMDk3MTQ2MDItLjIzNDAzMzYtLjMyNjE3MDMtLjM4NjA4NDgtLjU3OTU2Mi0uMzg0Nzc2MjV6Ii8+PC9nPjwvc3ZnPg=='),
    linear-gradient(transparent, transparent);
  background-repeat: no-repeat;
  appearance: none;
  background-color: white;
  background-position: right 0.75rem center;
  background-size: 0.5rem;
  padding-right: 2rem;

  &::-ms-expand {
    display: none;
  }
  &:-webkit-autofill {
    -webkit-appearance: menulist;
    appearance: menulist;
  }
  &:focus {
    border-color: #aaa;
    box-shadow: 0 0 1px 3px rgba(59, 153, 252, 0.7);
    box-shadow: 0 0 0 3px -moz-mac-focusring;
    color: #222;
    outline: none;
  }

  option {
    font-weight: normal;
  }
}
