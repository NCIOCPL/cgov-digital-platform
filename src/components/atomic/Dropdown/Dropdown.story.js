import React from 'react';
import generateStories from '../../../../.storybook/generateStories';
import Dropdown from './Dropdown';

generateStories({
  component: { Dropdown },
  storyName: 'Elements | Dropdown',
  defaultProps: {
    id: 'test-id',
    label: 'Example Dropdown Select Input',
    required: false,
    hasError: false,
    errorMessage: null,
    value: 'example-dropdown',
  },
  defaultChildren: (
    <>
      <option value="value1">Option A</option>
      <option value="value2">Option B</option>
      <option value="value3">Option C</option>
    </>
  ),
  stories: [
    {
      name: 'default',
    },
  ],
});
