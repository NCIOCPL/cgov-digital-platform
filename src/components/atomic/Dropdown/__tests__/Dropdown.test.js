import React from 'react';
import { mount } from 'enzyme';

import Dropdown from '../Dropdown.jsx';
jest.unmock('../Dropdown.jsx');

describe('Dropdown', () => {

  let wrapper = null;
  const labelText = 'label goes here';
  
  beforeEach(function() {
    wrapper = mount(
      <Dropdown label={labelText}>
        <option value='value1'>Option A</option>
        <option value='value2'>Option B</option>
        <option value='value3'>Option C</option>
      </Dropdown>);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the label text correctly', () => {
    const labelElement = wrapper.find('label');
    expect(labelElement.text()).toBe(labelText);
  });

  it('has an id attribute', () => {
    expect(typeof wrapper.find('select').props().id).toBe('string');
  });

  it('has a matching id and htmlFor attributes', () => {
    const id = wrapper.find('select').props().id;
    const htmlFor = wrapper.find('label').props().htmlFor;
    expect(htmlFor).toBe(id);
  });

  it('adds a placeholder if there\'s no value', () => {
    expect(wrapper.find('option').first().text()).toBe('Select ...');
  });

  it('accepts a value property', () => {
    const testValue = 'testing';
    wrapper = mount(<Dropdown value={testValue} label={labelText}>
      <option value='value1'>Option A</option>
      <option value='value2'>Option B</option>
      <option value='value3'>Option C</option>
    </Dropdown>);
    const value = wrapper.find('select').props().value;
    expect(value).toBe(testValue);
  });

  it('accepts the required property', () => {
    wrapper = mount(<Dropdown required label={labelText}>
      <option value='value1'>Option A</option>
      <option value='value2'>Option B</option>
      <option value='value3'>Option C</option>
    </Dropdown>);
    expect(wrapper.find('label').hasClass('usa-label--required')).toBe(true);
  });

  it('displays the error message when present', () => {
    const message = 'error message';
    wrapper = mount(<Dropdown errorMessage={message} label={labelText}>
      <option value='value1'>Option A</option>
      <option value='value2'>Option B</option>
      <option value='value3'>Option C</option>
    </Dropdown>);
    expect(wrapper.find('.usa-error-message').text()).toBe(message);
  });

  it('handles onChange event', () => {
    const testValue = 'value3';
    wrapper = mount(<Dropdown value={'value1'} label={labelText}>
      <option value='value1'>Option A</option>
      <option value='value2'>Option B</option>
      <option value='value3'>Option C</option>
    </Dropdown>);
    wrapper.find('select').simulate('change', {target: { value : testValue}});
    expect(wrapper.find('select').props().value).toBe(testValue);
  });
});
