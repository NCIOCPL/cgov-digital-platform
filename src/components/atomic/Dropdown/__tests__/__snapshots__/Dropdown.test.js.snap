// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render correctly 1`] = `
<Dropdown
  action={[Function]}
  hasError={false}
  label="label goes here"
  required={false}
>
  <InputLabel
    htmlFor="component-unique-id-0"
    label="label goes here"
    required={false}
  >
    <label
      className="undefined "
      htmlFor="component-unique-id-0"
      id="component-unique-id-0-label"
    >
      label goes here
    </label>
  </InputLabel>
  <select
    className="cts-select"
    id="component-unique-id-0"
    onChange={[Function]}
    required={false}
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      Select ...
    </option>
    <option
      value="value1"
    >
      Option A
    </option>
    <option
      value="value2"
    >
      Option B
    </option>
    <option
      value="value3"
    >
      Option C
    </option>
  </select>
</Dropdown>
`;
